room: refactor lava and trigger interpretation

lahm86 opened the pull request:
Resolves #941.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This completes the floor data migration into properties in `SECTOR_INFO`. The raw `int16_t` array from the level files is now only used temporarily while the level is loaded; once all FD is parsed, it is discarded. This has allowed `g_FloorData` to be removed.

We do some forward lookup when reading `TRIGGER` data to know how many `TRIGGER_CMD`s are needed to be stored. The benefit is that in retrieving data later in the codebase, we just need simple `for` loops rather than re-interpreting FD values repeatedly.

`Room_TestTriggers` now expects the "triggerer" (`ITEM_INFO`) to be passed to it and it works out the sector, "heavy" state and suchlike from there. This greatly simplifies calls to it, as we don't need to have guaranteed that `g_TriggerIndex` has been set via `Room_GetHeight`. That global - and the related pointer in `COLL_INFO` - have also been removed as they are no longer required.

I did wonder if we could add function pointers to `TRIGGER_CMD` to tidy the switch statement in `Room_TestTriggers` e.g. when parsing, we allocate the function pointer based on the command type. But this probably needs some further refactoring, such as making a new module, plus we need to track flip status, flip effect and suchlike throughout the switch statement, so it maybe needs some further thought.

One other niggle is that `index` in `SECTOR_INFO` is only used in the initial parsing. I can't think of a good way to store these temporarily and then discard, but this is maybe OK for now.

For floor data injection, no `.bin` changes are needed, despite the version being updated. This is only so that subsequent updates don't need to have a global FD injection size. We no longer append to `g_FloorData` during injection; instead, because it now takes place after sector/FD parsing, we just update the existing properties or have `room.c` parse the provided injection data where new entries are needed.

lahm86 committed:
room: refactor lava and trigger interpretation

This creates an is_death_sector property on sectors and updates the
checks for when this is applicable in terms of collision and trigger
checks.
Triggers are also now directly in sectors, and their list of commands.
Most duplicated FD iterations are now removed; injection logic to
follow suit and use the new approach.

lahm86 committed:
door: set NO_ROOM on blocked wall portals

This is effectively what setting the index was previously doing, the
new approach is now adopted.

lahm86 committed:
inject: refactor floor data injection

Floor data injection now uses the common parser in `room.c`. This means
making the sector populator public and passing additional arguments to
allow parsing from index 0.

lahm86 committed:
level: store raw floor data temporarily

This eliminates the global floor data array and instead reads it into
the temporary level info; it is released post-parsing.
Injections updated to match, as their total FD injected length is no
longer required.

lahm86 committed:
room: delete global trigger index

The global trigger index is no longer required, and as such neither is
the trigger pointer in COLL_INFO.

lahm86 committed:
inject: fix missing null check

walkawayy commented on src/game/room.c:
```diff
@@ -752,68 +730,65 @@ static void Room_PopulateSectorData(
         case TT_PAD:
         case TT_ANTIPAD:
-            if (g_LaraItem->pos.y != g_LaraItem->floor) {
+            if (item->pos.y != item->floor) {
```
Could this have an implication where an item other than Lara could trigger an antipad?

lahm86 replied:
That'd make for some interesting skips for sure 😄 
Other items will enter the `is_heavy` condition above (line 758) so it's only ever Lara in this `else` part 😌 I just thought a couple of references fewer to a global was nice where possible.

walkawayy replied:
Oh nice. I recently saw a playthrough of a custom level and the person tried to trigger a floor pad with a dead animal. Kinda genius but it didn't work. :laughing: 

So could we change `Lava_Burn(g_LaraItem);` to `Lava_Burn(item);` just to get rid of the global in that whole function?

lahm86 replied:
Oh yes. Although I think we'll have to cast to non const.

walkawayy reviewed (approved):
This looks amazing!!

After all the refactors, we should make sure testers like @aredfan and us do some proper play testing to make sure there are no regressions.

rr- reviewed (commented):
Call me overzealous, but what do you think about making TRIGGER.parameter a `void*`, and store the `camera` member in a linear GameBuffer to which we point then through `parameter`?

rr- commented on src/game/objects/general/door.c:
```diff
@@ -51,12 +51,12 @@ static void Door_Shut(DOORPOS_DATA *const d)
         return;
     }
 
-    sector->index = 0;
```
Can you explain a bit more regarding why this assignment is no longer necessary?

lahm86 replied:
This goes hand in hand with the line below.

rr- commented on src/game/objects/general/door.c:
```diff
@@ -51,12 +51,12 @@ static void Door_Shut(DOORPOS_DATA *const d)
     sector->ceiling.height = NO_HEIGHT;
     sector->portal_room.sky = NO_ROOM;
     sector->portal_room.pit = NO_ROOM;
+    sector->portal_room.wall = NO_ROOM;
```
Is it safe to introduce this assignment?

lahm86 replied:
Ah, apologies, I forgot to discuss this above. This was the last reference to the sector's index, along with in `Room_GetSector`, and I overlooked it in the last PR. When we had `Room_GetDoor`, setting the sector's index to 0 meant that that returned `NO_ROOM` when the door is shut. So this is just doing the same but more explicitly. When the door is opened, the original sector gets restored.
I think this "FD nulling" is in place otherwise Lara could glitch through closed doors and then `Room_GetSector` would push her through the portal.

lahm86 commented:
> Call me overzealous, but what do you think about making TRIGGER.parameter a `void*`, and store the `camera` member in a linear GameBuffer to which we point then through `parameter`?

Nice idea. I will try it out tomorrow.

lahm86 committed:
room: make trigger command parameter generic

Rather than storing redundant camera data in every command, we now
store a void *, which either has the regular int16_t parameter, or the
full additional camera data.

lahm86 commented:
I've changed `parameter` to `void *` in the command struct, and although there is a little extra casting needed here and there, I think overall it's much better. We're not allocating redundant space for camera data in every command crucially.
